### ISSUE TITLE
communication between 2 nodes

### ISSUE THREAD

**Issue by NicolasDP** [input-output-hk/jormungandr#612]
TL;DR:

<img src="https://user-images.githubusercontent.com/5445840/60728424-8df6dd00-9f40-11e9-8d5f-aa4fb73acd26.png" width="300">

# draw 2 nodes

the first part to work on is the minimal element that allows block synchronisation between nodes in the most naive way. Here it just needs to work and does not need to be pretty or optimised.

We don't need to care much about the transaction propagation because we still have end points we can send transaction to in the REST API. So working in BFT mode, we could open a couple of end points to allow HTTP POST to send transactions.

We also can ignore the P2P components yet. Even though the gossiping is working (ish) as it is we can ignore none critical errors as we will rely mainly on the trusted peers part for this milestone.

In order to reach block synchronisation between nodes we can work with what is known to work. We know we have the capability to synchronise all blocks **from** a given block id toward the peer's TIP (the last block of the peer).

- [x] when receiving a block header notification from a peer: request all blocks from a check point to synchronise the blocks with this peer. We might already re-download blocks we already have but this is fine for version 1.

having this in place will allow us to quickly have something that _works_ we can use between the developers to tests and develop improvements in the following milestones.

# draw the rest of the network

Following the work above, we can now start improving the blockchain synchronisation with other peers as well as starting improve other aspect of the network.

## first thing we know of

- [x] the `NodeId` is not needed as an object within the poldercast module. It is required by the paper but we can use an internal/arbitrary function to determine _proximity_ via hashing the peer's address or anything else; We can however keep an `Id` internally that we randomly generate to address a given peer within the node's internal processes; #757 
- [x] the block synchronisation can be improved largely: we need to be able to synchronise the blocks we need and not re-download a large part of the blockchain from another peer; For this reason we will need to store block headers in quarantine and start downloading intermediate block headers for validation. Then we can download the blocks for full validation and storage. #704 #1078 
- [x] we will need to add some sort of blacklisting capabilities for nodes that sends invalid blocks (we will need to start thinking on some policy settings but we can start simple at the beginning); #1058 

## make it pretty

- [x] improve the backlisting policy settings;
- [x] allow better concurrent block query in the _blockchain_ service/task;
- [x] rely more on the p2p topology (if not already the case);

## make it fast

- [ ] see how we can download part of the blockchain from multiple peers at the same time (bootstrap phase) or download epoch per epoch?
- [x] finalise the transaction support across the node;

**Comment by mzabaluev** [comment]
With two-stage download (`PullHeaders` followed by one or more `GetBlocks`) I got [to a point](https://github.com/input-output-hk/jormungandr/pull/626#issuecomment-510174688) where either branch quarantine support needs to be implemented in storage, or the way jormungandr uses the protocol should be simplified for expediency.

The simplest viable implementation is to request `PullBlocksToTip` from the node that announced a disconnected block, essentially doing a mini-bootstrap.

**Comment by NicolasDP** [comment]
Closing this issue as this is mostly done. We are now having nodes connecting fully as required by this call.
